mantidproject/mantid
# Change PlotAsymmetryByLogValue To Accept Range Of Runs As Input

**Description of work.**
Added a new array property to PlotAsymmetryByLogValue called WorkspaceNames. Changed the algorithm so that FirstRun and LastRun are no longer mandatory. For successful execution either WorksapceNames must be supplied, or FirstRun and LastRun must be supplied. Added unit tests to test this. 



**To test:**
Run below script and check results are as expected. Results using FirstRun and LastRun as input should be compared to using WorkspaceNames as input, should give the same output. Notice ws_list has only two runs so the output should only be from the first and last run, excluding the middle run. Whereas ws_list_full should be equivalent to using FirstRun and LastRun.
```
first_run = 'MUSR00015189.nxs'
middle_run = 'MUSR00015190.nxs'
last_run = 'MUSR00015191.nxs'

print('####################')
print('Example - Calculating asymmetry for a series of MUSR runs:')
print('####################')

print('First and Last')
ws = PlotAsymmetryByLogValue(FirstRun = first_run,
                             LastRun = last_run,
                             LogValue = 'sample_magn_field',
                             TimeMin = 0.55,
                             TimeMax = 12.0);

print("Y values (asymmetry): {}".format(ws.readY(0)))
print("X values (sample magn. field): {}".format(ws.readX(0)))

ws_list = [first_run, last_run]

print('Short List')
ws = PlotAsymmetryByLogValue(WorkspaceNames = ws_list,
                             LogValue = 'sample_magn_field',
                             TimeMin = 0.55,
                             TimeMax = 12.0);

print("Y values (asymmetry): {}".format(ws.readY(0)))
print("X values (sample magn. field): {}".format(ws.readX(0)))

ws_list_full = [first_run, middle_run, last_run]

print('Full List')
ws = PlotAsymmetryByLogValue(WorkspaceNames = ws_list_full,
                             LogValue = 'sample_magn_field',
                             TimeMin = 0.55,
                             TimeMax = 12.0);

print("Y values (asymmetry): {}".format(ws.readY(0)))
print("X values (sample magn. field): {}".format(ws.readX(0)))

print('####################')
print('Using both Red and Green periods:')
print('####################')

print('First and Last')
ws = PlotAsymmetryByLogValue(FirstRun = 'MUSR00015189.nxs',
                             LastRun = 'MUSR00015191.nxs',
                             LogValue = 'sample_magn_field',
                             TimeMin = 0.55,
                             TimeMax = 12.0,
                             Red = 1,
                             Green = 2);

print("Y values (difference): {}".format(ws.readY(0)))
print("Y values (red): {}".format(ws.readY(1)))
print("Y values (green): {}".format(ws.readY(2)))
print("Y values (sum): {}".format(ws.readY(3)))
print("X values (sample magn. field): {}".format(ws.readX(0)))

print('Short List')
ws = PlotAsymmetryByLogValue(WorkspaceNames = ws_list,
                             LogValue = 'sample_magn_field',
                             TimeMin = 0.55,
                             TimeMax = 12.0,
                             Red = 1,
                             Green = 2);

print("Y values (difference): {}".format(ws.readY(0)))
print("Y values (red): {}".format(ws.readY(1)))
print("Y values (green): {}".format(ws.readY(2)))
print("Y values (sum): {}".format(ws.readY(3)))
print("X values (sample magn. field): {}".format(ws.readX(0)))

print('Full List')
ws = PlotAsymmetryByLogValue(WorkspaceNames = ws_list_full,
                             LogValue = 'sample_magn_field',
                             TimeMin = 0.55,
                             TimeMax = 12.0,
                             Red = 1,
                             Green = 2);

print("Y values (difference): {}".format(ws.readY(0)))
print("Y values (red): {}".format(ws.readY(1)))
print("Y values (green): {}".format(ws.readY(2)))
print("Y values (sum): {}".format(ws.readY(3)))
print("X values (sample magn. field): {}".format(ws.readX(0)))
```

**To test the dialog window:** 

1. Set your default instrument to **MUSR**
2. Open the dialog window for **PlotAsymmetryByLogValue** from the algorithms widget on mantid workbench
3. In the runs box enter `15189-15191`. This will load files **MUSR15189, MUSR15190, MUSR15191**
4. Remember to enter a name for the **Output workspace**
5. Show data on the output workspace and hopefully you have something that looks as follows:

![image](https://user-images.githubusercontent.com/46685490/91427056-45416800-e855-11ea-88d4-f8257d579d52.png)

6. Feel free to test different ranges to check no unexpected errors occur, use commas to include individual runs e.g. `15189-15191,15195`

Fixes #29160 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
